### PR TITLE
feat(eap): Add trimming processor for EAP items

### PR DESCRIPTION
**Pull request by loewenheim**
This adds a new trimming processor specifically for EAP items. The reason we can't just reuse the existing processor is the size calculation logic. The existing processor calculates the sizes of values using a special serializer. For EAP items, specfically their attributes, we want to use the attribute size logic defined in `eap::size`.

Attributes are processed in order from smallest to greatest total (key + value) size. The effect of this is that small attributes are more likely to be preserved. Attribute keys are taken into account for the size, but never trimmed—if a key is too long the attribute is just removed.

ref: #5362. ref: RELAY-177.

### CODE REVIEW

File: relay-event-normalization/src/eap/trimming.rs [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        &mut self,
+        _value: Option<&T>,
+        _: &mut Meta,
+        state: &ProcessingState<'_>,
+    ) -> ProcessingResult {
+        // If our current depth is the one where we found a bag_size attribute, this means we
+        // are done processing a databag. Pop the bag size state.
+        self.size_state
+            .pop_if(|size_state| state.depth() == size_state.encountered_at_depth);
+
```
[the remark this answers is not in the document]
Comment (loewenheim) [reply relay-event-normalization/src/eap/trimming.rs:137]: The method has existed since 1.86.0 https://doc.rust-lang.org/std/vec/struct.Vec.html#method.pop_if.

File: relay-event-normalization/src/eap/trimming.rs [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+impl TrimmingProcessor {
+    /// Creates a new trimming processor.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    fn should_remove_container<T: Empty>(&self, value: &T, state: &ProcessingState<'_>) -> bool {
+        // Heuristic to avoid trimming a value like `[1, 1, 1, 1, ...]` into `[null, null, null,
+        // null, ...]`, making it take up more space.
+        self.remaining_depth(state) == Some(1) && !value.is_empty()
```
Comment (Dav1dde) [inline relay-event-normalization/src/eap/trimming.rs:62]: Do you have an example/test for this? We might need the `null` cases to individually attach metadata to each of the nodes, deleting the parent will delete the child metadata (which maybe is okay).

**Review by Dav1dde — APPROVED** [review]
Seems a good base to add more features like special keys, adding in support for conventions later.

If you can think of more examples and edge cases to write tests for, I think it would be a good time to do that now, these snapshots will come in very handy for every future change.

File: relay-event-normalization/src/eap/trimming.rs (hunk 2 above)
Comment (loewenheim) [reply relay-event-normalization/src/eap/trimming.rs:62]: I copied that for parity with the other trimming processor, but that doesn't seem to have a test for it either :/

**Comment by loewenheim** [comment]
>Seems a good base to add more features like special keys, adding in support for conventions later.
>
>If you can think of more examples and edge cases to write tests for, I think it would be a good time to do that now, these snapshots will come in very handy for every future change.

What I'd like to test is "dynamic" max chars/max bytes (similar to dynamic PII, i.e. being able to compute it with a function). Implementing that feature isn't difficult; now that I've done it for PII I think it should be very similar. I just can't think of how to mock it for testing purposes, since this processor's logic is directly tied to `Attributes`/`AttributeValue`s.

**Comment by Dav1dde** [comment]
@loewenheim sounds good, I like if we split the work and tackle the modifications afterwards and just merge this as is. I think for these future changes we'll benefit from a more extensive testsuite, so maybe adding more (snapshot) tests now will pay off soon, but obviously that's up to you.

:shipit:

File: relay-event-normalization/src/eap/trimming.rs [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+            self.consume_size(key.len());
+
+            if self.remaining_size() == Some(0) {
+                split_idx = Some(idx);
+                break;
+            }
+
+            let value_state = state.enter_borrowed(key, None, ValueType::for_field(value));
+            processor::process_value(value, self, &value_state)?;
+            self.consume_size(size::attribute_size(value))
```
[the remark this answers is not in the document]
Comment (loewenheim) [reply relay-event-normalization/src/eap/trimming.rs]: This isn't even the function's last line.

**Comment by loewenheim** [comment]
As written, the processor currently "overaccepts" numerical/boolean attributes. This means that if e.g. a `u64` attribute value would cause the size limit to be exceeded, we accept that attribute anyway and only start discarding attributes starting with the next one (the `test_overaccept_number` test demonstrates this). I believe this is acceptable because
1. it can cause us to overaccept 7B at maximum, so the effect is bounded
2. attribute key lengths don't vary _that_ widely; I would expect attribute sizes to be determined mostly by string lengths. As a consequence I would expect it to be very unlikely for a numerical/boolean attribute to be the cutoff in the first place.

@Dav1dde does that make sense?

File: relay-event-normalization/src/eap/trimming.rs [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+        let mut sorted: Vec<_> = inner.into_iter().collect();
+        sorted.sort_by_cached_key(|(k, v)| k.len() + size::attribute_size(v));
+
+        let mut split_idx = None;
+        for (idx, (key, value)) in sorted.iter_mut().enumerate() {
+            self.consume_size(key.len());
+
+            if self.remaining_size() == Some(0) {
+                split_idx = Some(idx);
+                break;
```
[the remark this answers is not in the document]
Comment (loewenheim) [reply relay-event-normalization/src/eap/trimming.rs]: Not even close to the correct line. Also, no, `pop_if` is not unstable.

File: relay-event-normalization/src/eap/trimming.rs [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+        attributes.0 = sorted.into_iter().collect();
+
+        let new_size = size::attributes_size(attributes);
+        if new_size != original_length {
+            meta.set_original_length(Some(original_length));
+        }
+
+        self.in_attributes = false;
+
+        Ok(())
```
[the remark this answers is not in the document]
Comment (loewenheim) [reply relay-event-normalization/src/eap/trimming.rs:358]: Fixed.

File: relay-event-normalization/src/eap/trimming.rs [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+                ],
+                "len": 27
+              }
+            }
+          }
+        }
+        "###);
+    }
+
+    fn test_overaccept_number() {
```
[the remark this answers is not in the document]
Comment (loewenheim) [reply relay-event-normalization/src/eap/trimming.rs:529]: Fixed.

**Review by jjbayer — COMMENTED** [review]
I'd like to revert my opinion on whether this should be a separate processor:

1. The two trimming processors now share a lot of copy-pasted code.
2. The difference seems to be mainly in `process_attributes`, which the original trimming processor does not override.

Alternatively, we could deduplicate code by embedding a `trimming::TrimmingProcessor` within the new one, and delegating to it.

**Comment by loewenheim** [comment]
>Alternatively, we could deduplicate code by embedding a trimming::TrimmingProcessor within the new one, and delegating to it.

I don't think that would work because of the size calculation logic.

I agree that the duplication is super unfortunate, but I don't see how we can avoid it without making it possible to generalize size calculations.

**Comment by jjbayer** [comment]
> I don't think that would work because of the size calculation logic.
> 
> I agree that the duplication is super unfortunate, but I don't see how we can avoid it without making it possible to generalize size calculations.

Alright, not a blocker so feel free to merge.

File: relay-event-normalization/src/eap/mod.rs [hunk 7]
```diff
@@ -24,6 +24,8 @@ use crate::{ClientHints, FromUserAgentInfo as _, RawUserAgentInfo};
 
 mod ai;
 mod size;
+#[allow(unused)]
```
Comment (jjbayer) [inline relay-event-normalization/src/eap/mod.rs:27]:
```suggestion
#[expect(unused)]
```

File: relay-event-normalization/src/eap/trimming.rs [hunk 8]
```diff
@@ -0,0 +1,767 @@
...
+    }
+
+    #[inline]
+    fn remaining_depth(&self, state: &ProcessingState<'_>) -> Option<usize> {
+        self.size_state
+            .iter()
+            .filter_map(|size_state| {
+                // The current depth in the entire event payload minus the depth at which we found the
+                // max_depth attribute is the depth where we are at in the property.
+                let current_depth = state.depth() - size_state.encountered_at_depth;
```
Comment (jjbayer) [inline relay-event-normalization/src/eap/trimming.rs:80]: Can this ever overflow?

File: relay-event-normalization/src/eap/trimming.rs [hunk 9]
```diff
@@ -0,0 +1,767 @@
...
+        Ok(())
+    }
+    fn process_u64(
+        &mut self,
+        _value: &mut u64,
+        _meta: &mut Meta,
+        _state: &ProcessingState<'_>,
+    ) -> ProcessingResult {
+        if self.in_attributes {
+            self.consume_size(8);
```
Comment (jjbayer) [inline relay-event-normalization/src/eap/trimming.rs:151]: nit for this and the functions below.

```suggestion
            self.consume_size(size_of_val(value));
```

File: relay-event-normalization/src/eap/trimming.rs [hunk 10]
```diff
@@ -0,0 +1,767 @@
...
+            crate::trimming::trim_string(value, meta, max_chars, state.attrs().max_chars_allowance);
+        }
+
+        if !state.attrs().trim {
+            self.consume_size(value.len());
+            return Ok(());
+        }
+
+        if let Some(size_remaining) = self.remaining_size() {
+            crate::trimming::trim_string(value, meta, size_remaining, 0);
```
Comment (jjbayer) [inline relay-event-normalization/src/eap/trimming.rs:208]: What happens when the string is longer than or equal to the remaining size? Shouldn't the function return `DeleteValueHard` in that case?

File: relay-event-normalization/src/eap/mod.rs (hunk 7 above)
Comment (loewenheim) [reply relay-event-normalization/src/eap/mod.rs:27]: I could've sworn I tried that and it didn't work because of tests, but apparently I'm wrong.

File: relay-event-normalization/src/eap/trimming.rs (hunk 8 above)
Comment (loewenheim) [reply relay-event-normalization/src/eap/trimming.rs:80]: Logically, no. I'm still going to replace it with a safe alternative, thanks for catching it.

File: relay-event-normalization/src/eap/trimming.rs (hunk 9 above)
Comment (loewenheim) [reply relay-event-normalization/src/eap/trimming.rs:151]: The 1/8 are the sizes defined in the `eap::size` module. Maybe they should be exported as constants?

File: relay-event-normalization/src/eap/trimming.rs (hunk 10 above)
Comment (loewenheim) [reply relay-event-normalization/src/eap/trimming.rs:208]: How do you mean? The purpose of `trim_string` is exactly to cut down the string to fit into the remaining size. Do you mean what happens if that is not possible?

Comment (jjbayer) [reply relay-event-normalization/src/eap/trimming.rs:208]: What `trim_string` does make sense to me, but if the string is empty after calling `trim_string`, should `process_string` return an error? Or do we just leave the empty string in place?

Comment (loewenheim) [reply relay-event-normalization/src/eap/trimming.rs:208]: If we have 0 size left before we even start trimming the string, it's already filtered out in `before_process`.

**Review by jjbayer — APPROVED** [review]
